plugin/src/main/scala/tut/TutPlugin.scala

```diff
@@ -8,65 +8,31 @@ import sbt.Attributed.data
-
-  def safeListFiles(dir: File, recurse: Boolean): List[File] =
-    Option(dir.listFiles).fold(List.empty[File]){ files =>
-      val l = files.toList
-      if (recurse) l.flatMap(flatten) else l
-    }
-
-  def safeRelativize(in: File, out: File, recurse: Boolean): List[(File, String)] = {
-    val inPath = in.toPath
-    val outPath = out.toPath
-    val read = safeListFiles(in, recurse = true).map(f => inPath.relativize(f.toPath)).toSet
-    safeListFiles(out, recurse = true).flatMap { f =>
-      val rel = outPath.relativize(f.toPath)
-      if (read(rel)) (f -> rel.toString) :: Nil else Nil
-    }
-  }
+  override def trigger = allRequirements
 
-  def flatten(f: File): List[File] =
-    f :: (if (f.isDirectory) f.listFiles.toList.flatMap(flatten) else Nil)
+  override def requires = sbt.plugins.JvmPlugin
 
-  def tutFilesParser(state: State): Parser[File] = {
-    val extracted = Project.extract(state)
-    val dir     = extracted.getOpt(tutSourceDirectory)
-    val files   = dir.fold(List.empty[File])(d => safeListFiles(d, recurse = true))
-    val parsers = dir.fold(List.empty[Parser[File]])(dir => files.map(f => literal(dir.toURI.relativize(f.toURI).getPath).map(_ => f)))
-    val folded  = parsers.foldRight[Parser[File]](failure("<no input files>"))(_ | _)
-    token(folded)
-  }
+  type Dir = File
 
-  /** Run the Tut CLI for a single input file or directory */
-  def tutOne(streams: TaskStreams, r: ScalaRun, in: File, out: File, cp: Classpath, opts: Seq[String], pOpts: Seq[String], re: String): List[(File, String)] = {
-    toError(r.run("tut.TutMain",
-      data(cp),
-      Seq(in.getAbsolutePath, out.getAbsolutePath, re) ++ opts ++ pOpts,
-      streams.log))
-    // We can't return a value from the runner, but we know what TutMain is looking at so we'll
-    // fake it here. Returning all files potentially touched.
-    safeRelativize(in, out, recurse = true)
+  object autoImport {
+    lazy val Tut                = config("tut") extend Compile
+    lazy val tut                = TaskKey[Seq[(File,String)]]("tut", "create tut documentation")
+    lazy val tutSourceDirectory = SettingKey[File]("tutSourceDirectory", "where to look for tut sources")
+    lazy val tutPluginJars      = TaskKey[Seq[File]]("tutPluginJars", "Plugin jars to be used by tut REPL.")
+    lazy val tutOnly            = inputKey[Unit]("Run tut on a single file.")
+    lazy val tutTargetDirectory = SettingKey[File]("tutTargetDirectory", "Where tut output goes")
+    lazy val tutNameFilter      = SettingKey[Regex]("tutNameFilter", "tut skips files whose names don't match")
+    lazy val tutFiles           = SettingKey[State => Parser[File]]("tutFiles", "parser identifying files visible to tut")
+    lazy val tutQuick           = TaskKey[Set[File]]("tutQuick", "Run tut incrementally on recently changed files")
+    lazy val tutQuickCache      = SettingKey[File]("tutQuickCache", "Cache directory for tutQuick metadata")
```

Comment:
I'd prefer macros like `settingKey[File]("Cache directory for tutQuick metadata")` here for consistency and to prevent mistake.